plugin/trino-hive/src/main/java/io/trino/plugin/hive/fs/TransactionScopeCachingDirectoryLister.java

```diff
@@ -0,0 +1,230 @@
+                    cache.asMap().replace(path, cachedValueHolder, cachedValueHolder);
+                    return hasNext;
+                }
+                catch (IOException exception) {
+                    // invalidate cached value to force retry of directory listing
+                    cache.invalidate(path);
+                    throw exception;
+                }
+            }
+
+            @Override
+            public LocatedFileStatus next()
+                    throws IOException
+            {
+                // force cache entry weight update in case next file is cached
+                checkState(hasNext());
+                return cachedValueHolder.getCachedFile(fileIndex++).orElseThrow();
+            }
+        };
+    }
+
+    @VisibleForTesting
+    boolean isCached(Path path)
+    {
+        FetchingValueHolder cached = cache.getIfPresent(path);
+        return cached != null && cached.isFullyCached();
+    }
+
+    private static boolean isLocationPresent(Storage storage)
+    {
+        // Some Hive table types (e.g.: views) do not have a storage location
+        return storage.getOptionalLocation().isPresent() && isNotEmpty(storage.getLocation());
+    }
+
+    private static class FetchingValueHolder
+    {
+        private final List<LocatedFileStatus> cachedFiles = synchronizedList(new ArrayList<>());
+        @GuardedBy("this")
+        @Nullable
+        private RemoteIterator<LocatedFileStatus> fileIterator;
+        @GuardedBy("this")
+        @Nullable
+        private IOException exception;
+
+        public FetchingValueHolder(RemoteIterator<LocatedFileStatus> fileIterator)
+        {
+            this.fileIterator = requireNonNull(fileIterator, "fileIterator is null");
+        }
+
+        public synchronized boolean isFullyCached()
+        {
+            return fileIterator == null && exception == null;
+        }
+
+        public int getCachedFilesSize()
+        {
+            return cachedFiles.size();
+        }
+
+        public RemoteIterator<LocatedFileStatus> getCachedFiles()
+        {
+            checkState(isFullyCached());
+            return new SimpleRemoteIterator(cachedFiles.iterator());
+        }
+
+        public Optional<LocatedFileStatus> getCachedFile(int index)
+                throws IOException
+        {
+            int filesSize = cachedFiles.size();
+            checkArgument(index >= 0 && index <= filesSize);
+
+            if (index < filesSize) {
+                return Optional.of(cachedFiles.get(index));
+            }
+
+            return fetchNextCachedFile(index);
+        }
+
+        private synchronized Optional<LocatedFileStatus> fetchNextCachedFile(int index)
+                throws IOException
+        {
+            if (index < cachedFiles.size()) {
+                // file was fetched concurrently
+                return Optional.of(cachedFiles.get(index));
+            }
```

Comment:
Double check pattern doesn't except you from synchronization, just makes it less apparent -- the `cachedFiles` is synchronized after all.

just make `getCachedFile` synchronized, make the `cachedFiles` unsynchronized and `@GuardedBy`,
and then inline `fetchNextCachedFile`

Reply:
just made everything synchronized

Reply:
I actually restored two phase synchronization. This is because I would like to avoid synchronization in same section as `fileIterator.next()` (IO operation) for returning already cached files (which should be non-blocking)